google/cloud/pubsublite/internal/resumable_async_streaming_read_write_rpc.h

```diff
@@ -0,0 +1,469 @@
+      StreamInitializer<RequestType, ResponseType> initializer)
+      : retry_policy_prototype_(std::move(retry_policy)),
+        backoff_policy_prototype_(std::move(backoff_policy)),
+        sleeper_(std::move(sleeper)),
+        stream_factory_(std::move(stream_factory)),
+        initializer_(std::move(initializer)) {}
+
+  ResumableAsyncStreamingReadWriteRpcImpl(ResumableStreamingReadRpc&&) = delete;
+  ResumableAsyncStreamingReadWriteRpcImpl& operator=(
+      ResumableStreamingReadRpc&&) = delete;
+
+  future<Status> Start() override {
+    future<Status> status_future;
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      status_future = status_promise_.get_future();
+    }
+    Initialize(retry_policy_prototype_->clone(),
+               backoff_policy_prototype_->clone());
+    return status_future;
+  }
+
+  future<absl::optional<ResponseType>> Read() override {
+    future<absl::optional<ResponseType>> read_future;
+    bool retrying = false;
+
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      retrying = stream_state_ == State::kRetrying;
+      if (retrying) {
+        read_reinit_done_.emplace(absl::make_optional(promise<void>()));
+        read_future = read_reinit_done_->get_future();
+      } else {
+        read_future = stream_->Read();
+        outstanding_read_ = true;
+      }
+    }
+
+    if (retrying) {
+      return read_future;
+    }
+
+    return read_future.then(
+        [this](future<absl::optional<Response>> optional_response_future) {
+          auto optional_response = optional_response_future.get();
+          bool response_good = optional_response.has_value();
+          bool should_finish = false;
+
+          {
+            std::lock_guard<std::mutex> g{mu_};
+            outstanding_read_ = false;
+
+            // This means that an unsuccessful `Write` occurred, but could not
+            // enter the retry loop because of this present outstanding `Read`.
+            if (stream_state_ == State::kMustRetry) {
+              should_finish = true;
+              stream_state_ = State::kRetrying;
+            }
+            if (!response_good) {
+              read_reinit_done_.emplace(absl::make_optional(promise<void>()));
+
+              // If an outstanding write is present, we can't enter the retry
+              // loop, so we defer it until the outstanding `Write` finishes at
+              // which point we can enter the retry loop. Since we will return
+              // `read_reinit_done_`, we guarantee that another `Read` is not
+              // called while we're waiting for the outstanding write to finish
+              // and the retry loop to finish afterward.
+              if (outstanding_write_) {
+                stream_state_ = State::kMustRetry;
+              } else {
+                should_finish = true;
+                stream_state_ = State::kRetrying;
+              }
+            }
+          }
+
+          if (should_finish) {
```

Comment:
get the Finish logic out of the read and write calls, it should just be part of the retry logic